In [1]:
import music21
import os
import glob
import json
import csv
import pandas as pd

/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/urllib3/__init__.py:34: NotOpenSSLWarning: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


In [2]:
#Elaborate on all the filters, provide the results
base_dir1 = 'CPDL NO REPEAT_1'
base_dir2 = 'CPDL NO REPEAT_2'
all_score_files_dir1 = os.listdir(base_dir1)
all_score_files_dir2 = os.listdir(base_dir2)

In [3]:
len(all_score_files_dir1), len(all_score_files_dir2)

(353, 536)

## We find a total of 353 scores in directory1 and 536 scores in directory 2 i.e. a toal of 889 score files

### Score Filters Criteria

### (1) Filter all scores that are not parsable i.e. cannot be read using xml reader using music21

### Parse for directory 1

In [5]:
#Script to check the files that can be opened sucessfully using music21!
possible_file_type_extensions = ['*.xml', '*.musicxml']
total_correct_files = 0
correct_file_names = []
for score_file in all_score_files_dir1:
    if score_file=='.DS_Store':
        continue
    base_path = os.path.join(base_dir1, score_file)
    
    for file_extension in possible_file_type_extensions:
        musicxml_file = sorted(glob.glob(os.path.join(base_path, file_extension)))
        if len(musicxml_file)!= 0:
            xml_file_path = musicxml_file[0] 
    try:
        score = music21.converter.parse(xml_file_path)
        correct_file_names.append(xml_file_path)
    except:
        print("Could not parse", musicxml_file)

/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:5128: MusicXMLWarning: Could not import wedge: Error in getting DynamicWedges
  warnings.warn(f'Could not import {tag}: {excep}', MusicXMLWarning)


Could not parse ['CPDL NO REPEAT_1/EnquirerIngalls1805ax/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/134_de_ulenberg_cpdl/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/FirmamentRead1793bpr/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/FalmouthMaxim1802bpr/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Goodnight_from_the_Rhine_Op/score.musicxml']


/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 4 in part C..
  warnings.warn(


Could not parse ['CPDL NO REPEAT_1/HeadOfTheChurchIngalls1805ax/score.musicxml']


/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 12 in part C..
  warnings.warn(


Could not parse ['CPDL NO REPEAT_1/HarmonyJenks1800bpr/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/120_de_ulenberg_cpdl/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Hail_glorious_spirits_Tye/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/How_great_is_the_goodness/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Evening_Hiles/score.musicxml']


/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 2 in part Tr..
  warnings.warn(


Could not parse ['CPDL NO REPEAT_1/FluvannaWoodruff1800a/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/I_will_arise_Wood/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/JudeaHowe1804a/score.musicxml']


/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 1 in part Tr..
  warnings.warn(


Could not parse ['CPDL NO REPEAT_1/Gray_BoS_1_From_Fairest_SABpn/score.musicxml']


/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 4 in part B..
  warnings.warn(


Could not parse ['CPDL NO REPEAT_1/ImpartialSongIngalls1805ax/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Flowers_Zimmermann/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Good_Friday/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/GuilfordPilsbury1799bpr/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/He_who_ponders/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/I_saw_lovely_Phillis_Pearsall/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/HatfieldGillet1794bpr/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/HollandSwan1801bpr/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Humfrey-A_hymne_to_God_the_Father/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/HarmonyIngalls1805ax/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Five_lesser_joys_with_accomp_Warlock/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Gloria_in_F_Stainer/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Haydn_6_psalms_69_o_let_me_in/score.musicxml']
Could not pars

/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 8 in part B..
  warnings.warn(


Could not parse ['CPDL NO REPEAT_1/JubileeIngalls1805ax/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Hymn_to_Cynthia_Smart/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Evening_Service_in_G/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/GileadHolden1806a/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/11_descendid/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/How_sweet_the_answer_Parry/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Home_that_I_love_Abt/score.musicxml']


/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 16 in part T..
  warnings.warn(


Could not parse ['CPDL NO REPEAT_1/HephzipbahHusband1790a/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Haydn_6_psalms_41_maker_of_all/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Gibbons_Short_Service_in_F_Cantus_and_KBoard_Redux/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/For_anyone_who_drinks_-alpha/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/HallelujahHymnIngalls1805ax/score.musicxml']


/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 7 in part C..
  warnings.warn(


Could not parse ['CPDL NO REPEAT_1/FellowshipIngalls1805ax/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/26_Angelus_Domini/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/I_am_the_true_vine/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/GreatPhysicianIngalls1805bpr/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/HopkintonHolden1796bpr/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Feroci_Deficiente_Vino/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/18_Benedictus_es/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/Hunting_Song_Smart/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/13_delosalamos/score.musicxml']
Could not parse ['CPDL NO REPEAT_1/GiffW-HowNumerous/score.musicxml']


### Parse for directory 2

In [6]:
#PArse directory 2
possible_file_type_extensions = ['*.xml', '*.musicxml']
total_correct_files_dir2 = 0
correct_file_names_dir2 = []
for score_file in all_score_files_dir2:
    if score_file=='.DS_Store':
        continue
    base_path = os.path.join(base_dir2, score_file)
    for file_extension in possible_file_type_extensions:
        musicxml_file = sorted(glob.glob(os.path.join(base_path, file_extension)))
        if len(musicxml_file)!= 0:
            xml_file_path = musicxml_file[0] 
    try:
        score = music21.converter.parse(xml_file_path)
        print(xml_file_path)
        correct_file_names_dir2.append(xml_file_path)
    except:
        print("Could not parse", musicxml_file)

CPDL NO REPEAT_2/WainJ-OCome/score.musicxml
CPDL NO REPEAT_2/Vater_unser%2C_SWV_411_Schutz/score.musicxml
CPDL NO REPEAT_2/So_ell_encina_Anon/score.musicxml
CPDL NO REPEAT_2/VeniCreatorTallis1567bpr/score.musicxml
CPDL NO REPEAT_2/O_come%2C_let_us_sing_to_the_Lord_Tours/score.musicxml
CPDL NO REPEAT_2/Luna_Barnby/score.musicxml
CPDL NO REPEAT_2/MontvilleWoodruff1800bpr/score.musicxml
CPDL NO REPEAT_2/Ws-trad-cua/score.musicxml
CPDL NO REPEAT_2/MassJ-HeThatHasMade/score.musicxml
CPDL NO REPEAT_2/Tard-cum/score.musicxml
CPDL NO REPEAT_2/Nh_Schumann_Rotes_R%C3%B6sleint/score.musicxml
CPDL NO REPEAT_2/Ws-pal-ave1/score.musicxml
CPDL NO REPEAT_2/Lo%21_the_peaceful_shades_of_evening_Hatton/score.musicxml
CPDL NO REPEAT_2/The_Lord_anointed/score.musicxml
CPDL NO REPEAT_2/Vierne_-_Ave_Maria/score.musicxml
CPDL NO REPEAT_2/Wann_unsre_Augen_schlafe_ein_Schutz/score.musicxml
CPDL NO REPEAT_2/Veitch_I_heard_the_bells_on_Christmas_Day/score.musicxml
CPDL NO REPEAT_2/Look_toward_the_Lord/score.music

/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 10 in part Tr..
  warnings.warn(


CPDL NO REPEAT_2/MorganWSmith1803a/score.musicxml
CPDL NO REPEAT_2/Upton-Bread/score.musicxml
CPDL NO REPEAT_2/Nh_Gl%C3%A4ser_Lobet/score.musicxml
CPDL NO REPEAT_2/Verrijt-Ave_dulcis_Iesu/score.musicxml
CPDL NO REPEAT_2/WoO_74_Wanderers_Nachtlied-Rheinberger-A-Kl/score.musicxml
CPDL NO REPEAT_2/Nh_Vulpius_Nun_komm/score.musicxml
CPDL NO REPEAT_2/Love_not_me_for_comely_grace/score.musicxml
CPDL NO REPEAT_2/Lord_it_belongs_not_to_my_care_Davies/score.musicxml
CPDL NO REPEAT_2/O_Christ%2C_O_blessed_Lord_Wagner/score.musicxml
CPDL NO REPEAT_2/Love_me_little_Wilson/score.musicxml
CPDL NO REPEAT_2/Nh_Dorn_Fichtenbaum/score.musicxml
CPDL NO REPEAT_2/So_weich_und_warm_Cornelius/score.musicxml
CPDL NO REPEAT_2/The_Lord_is_my_light_Lawes/score.musicxml
CPDL NO REPEAT_2/Nh_Staden_Gelobet/score.musicxml
CPDL NO REPEAT_2/Moonlight_jf/score.musicxml
CPDL NO REPEAT_2/Thiele_Double_Chant_A_minor/score.musicxml
CPDL NO REPEAT_2/Now_fie_on_love_MacFarren/score.musicxml
CPDL NO REPEAT_2/Nh_Abt_Verlass_mi

/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 4 in part Tr..
  warnings.warn(


CPDL NO REPEAT_2/Land-ho_Leslie/score.musicxml
CPDL NO REPEAT_2/The_hemlock_tree_ATTB_Hatton/score.musicxml
CPDL NO REPEAT_2/Lassus_Agimus_tibi_gratias/score.musicxml
CPDL NO REPEAT_2/Victoria_asperges_me/score.musicxml
CPDL NO REPEAT_2/Lord%2C_I_have_loved_the_habitation_Torrance/score.musicxml
CPDL NO REPEAT_2/Ws-albe-tag/score.musicxml
CPDL NO REPEAT_2/My_days_are_gone_like_a_shadow_Blow/score.musicxml
CPDL NO REPEAT_2/Ws-stet-bea/score.musicxml
CPDL NO REPEAT_2/Weelkes_-_Most_mighty_and_all-knowing_Lord/score.musicxml
CPDL NO REPEAT_2/MentzHolyoke1797bpr/score.musicxml
CPDL NO REPEAT_2/O_Fools%21_can_you_not_see_orig/score.musicxml
CPDL NO REPEAT_2/Tollius-Audite%2C_stirps_Levitica/score.musicxml
CPDL NO REPEAT_2/The_unknown_land_tr_Hensel/score.musicxml
CPDL NO REPEAT_2/Tollius-Accipite_Spiritum_sanctum/score.musicxml
CPDL NO REPEAT_2/WhitWJ-ISendTheJoys-nokeyb/score.musicxml
CPDL NO REPEAT_2/NorwayShaw1815bpr/score.musicxml
CPDL NO REPEAT_2/Kuhnau-JS-Bach-Der-gerechte_Choeur/scor

CPDL NO REPEAT_2/The_East_Indian_Gregory/score.musicxml
CPDL NO REPEAT_2/Nh_Breitengraser_Sich_hat/score.musicxml
CPDL NO REPEAT_2/Todos_los_bienes_Encina/score.musicxml
CPDL NO REPEAT_2/LongS-HeThatHasGod/score.musicxml
CPDL NO REPEAT_2/Night_song_Abt/score.musicxml
CPDL NO REPEAT_2/Mh-Rot25-Partitur/score.musicxml
CPDL NO REPEAT_2/Ser-ben/score.musicxml
CPDL NO REPEAT_2/Ws-han-ch76/score.musicxml
CPDL NO REPEAT_2/Mh-HoeEng-Der_Engel_des_Herrn/score.musicxml
CPDL NO REPEAT_2/Lassus-Christus_resurgens_3v/score.musicxml
CPDL NO REPEAT_2/Weelkes-If_thy_deceitful_looks/score.musicxml
CPDL NO REPEAT_2/Tor-adm/score.musicxml
CPDL NO REPEAT_2/Ws-jos-ave6/score.musicxml
CPDL NO REPEAT_2/Tor-tie/score.musicxml
CPDL NO REPEAT_2/The_lifeboat_Hatton/score.musicxml
CPDL NO REPEAT_2/Nh_Feyhl_heilge/score.musicxml
CPDL NO REPEAT_2/O_give_thanks_Purcell/score.musicxml
CPDL NO REPEAT_2/Ojos_garcos_Vasquez/score.musicxml
CPDL NO REPEAT_2/Nh_Reger_VI_13_12/score.musicxml
CPDL NO REPEAT_2/Weelkes-A_count

/Users/jyotinarang/essentia_env/lib/python3.9/site-packages/music21/musicxml/xmlToM21.py:1986: MusicXMLWarning: The following exception took place in m. 1 in part C..
  warnings.warn(


CPDL NO REPEAT_2/MassJ-AriseAndHail/score.musicxml
CPDL NO REPEAT_2/NewOrleansBoyd1817bpr/score.musicxml
CPDL NO REPEAT_2/Unser_Herr_Jesus_Christus_in_der_Nacht_Schutz/score.musicxml
CPDL NO REPEAT_2/O_almighty_God/score.musicxml
CPDL NO REPEAT_2/Mh-Wannkom-Wann_kommest/score.musicxml
CPDL NO REPEAT_2/KnapiciusPsalm1/score.musicxml
CPDL NO REPEAT_2/Nunc_dimittis_in_D_Minor_Stainer/score.musicxml
CPDL NO REPEAT_2/Praetorius_-_Erhalt_uns_Herr/score.musicxml
CPDL NO REPEAT_2/WhelZ-HoEveryone/score.musicxml
CPDL NO REPEAT_2/Lassus-Ave_Regina_3v/score.musicxml
CPDL NO REPEAT_2/Vivaldi-Dixit_Dominus-RV_595-1.Dixit_Dominus/score.musicxml
CPDL NO REPEAT_2/WoO_82_Altdeutsches_Liebeslied-Rheinberger/score.musicxml
CPDL NO REPEAT_2/MeditationWest1807ax/score.musicxml
CPDL NO REPEAT_2/Make_me_a_clean_heart%2C_O_God_Barnby/score.musicxml
CPDL NO REPEAT_2/Let_my_complaint_come_before_thee_Batten/score.musicxml
CPDL NO REPEAT_2/Kings_and_Queens_Pinsuti/score.musicxml
CPDL NO REPEAT_2/No11-Aus_Der_Tie

CPDL NO REPEAT_2/WillAK-HowHappyIs-nokeyb/score.musicxml
CPDL NO REPEAT_2/White_Deus_Misereatur_Nostri/score.musicxml
CPDL NO REPEAT_2/Nh_Koschat_Verlassen/score.musicxml
CPDL NO REPEAT_2/Mastioletti_-_Terribilis_est_locus/score.musicxml
CPDL NO REPEAT_2/The_Lord_is_my_portion_Hopkins/score.musicxml
CPDL NO REPEAT_2/The_Lord_is_full_of_compassion_Boyce/score.musicxml
CPDL NO REPEAT_2/ThorW-TheLord-nokeyb/score.musicxml
CPDL NO REPEAT_2/Nh_M%C3%B6hring_Herr_wir_danken/score.musicxml
CPDL NO REPEAT_2/Nh_FescaEF_In_die_Ferne/score.musicxml
CPDL NO REPEAT_2/Tollius-Domine_quinque_talenta/score.musicxml
CPDL NO REPEAT_2/Nh_Klein_Der_Herr_ist_unsre_Zuversicht/score.musicxml
CPDL NO REPEAT_2/Lend_your_ears_to_my_sorrow/score.musicxml
CPDL NO REPEAT_2/O_Holy_Ghost%2C_into_our_minds_Macfarren_GA/score.musicxml
CPDL NO REPEAT_2/Morning_greeting_Hensel/score.musicxml
CPDL NO REPEAT_2/SmitJ-OAllYeNations/score.musicxml
CPDL NO REPEAT_2/Lullaby_of_Life_hl/score.musicxml
CPDL NO REPEAT_2/Tollius-All

# Results
### We obtain a total of 296 files from folder 1 and 531 files for folder 2 i.e. a total of 827 folders

## Back to more filters ...
### (2) Filter all scores for which the count of the number of streams in ‘voices’ folder is not the same as the number of streams in the xml file read using music21

## Parse for directory 1

In [9]:
#Of correct file names, get the number of scores that have the same count of files in the voices folder
correct_files_with_all_voices_dir1 = []
for music_xml_file in correct_file_names:
    score = music21.converter.parse(music_xml_file)
    # Get the number of streams in the parsed score
    num_streams = len(score.getElementsByClass(music21.stream.Stream))
    #print("Number of streams in the parsed score:", num_streams)
    #Test if the number of streams is the same as the number of voices in the voices folder
    voices_folder_path = os.path.join(os.path.dirname(music_xml_file), 'voices')
    all_voices = os.listdir(voices_folder_path)
    if len(all_voices)==num_streams:
        correct_files_with_all_voices_dir1.append(music_xml_file)

In [15]:
len(correct_files_with_all_voices_dir1)

147

## Parse for directory 2

In [12]:
#Script to extract streams of information with only 4 streams of all correct file names, group the scores by 
#count of streams
#Of correct file names, get the number of scores that have the same count of files in the voices folder
correct_files_with_all_voices_dir2 = []
for music_xml_file in correct_file_names_dir2:
    score = music21.converter.parse(music_xml_file)
    # Get the number of streams in the parsed score
    num_streams = len(score.getElementsByClass(music21.stream.Stream))
    #print("Number of streams in the parsed score:", num_streams)
    #Test if the number of streams is the same as the number of voices in the voices folder
    voices_folder_path = os.path.join(os.path.dirname(music_xml_file), 'voices')
    all_voices = os.listdir(voices_folder_path)
    if len(all_voices)==num_streams:
        correct_files_with_all_voices_dir2.append(music_xml_file)

In [14]:
len(correct_files_with_all_voices_dir2)

311

# Results
###  We obtain a total of len 147 files for directory 1 and 311 files from directory 2 i.e. a total of 458 files

## Back to more filters ...

## We now apply some musical criteria that was found by listening to several  songs from both the folders by a musician 

## (3) Filter out all files with 'Tr..mp3' in the voices folder - this criteria is specific to the dataset
#### Automatically discard all the folders that name the first voice (often named as Soprano or Cantus or Voice or Voice 1) “Treble” or “Tr”; they correspond to audio files “treble.mp3”, “tr..mp3”, “Tr.mp3” and for this DATASET they are instrumental generated instead of voice generated, even though they should be voice generated because they carry the leading voice


## Parse directory 1

In [26]:
#select the voices folder again and check for all files with Tr..mp3
correct_files_with_all_voices_dir1_filter_treble = []
for music_xml_file_path in correct_files_with_all_voices_dir1:
    score = music21.converter.parse(music_xml_file_path)
    # Get the number of streams in the parsed score
    num_streams = len(score.getElementsByClass(music21.stream.Stream))
    #print("Number of streams in the parsed score:", num_streams)
    #Test if the number of streams is the same as the number of voices in the voices folder
    voices_folder_path = os.path.join(os.path.dirname(music_xml_file_path), 'voices')
    all_voices = os.listdir(voices_folder_path)
    if 'Tr..mp3' in all_voices:
        continue
    else:
        correct_files_with_all_voices_dir1_filter_treble.append(music_xml_file_path)

In [27]:
len(correct_files_with_all_voices_dir1_filter_treble)

120

## Parse directory 2

In [28]:
#select the voices folder again and check for all files with Tr..mp3
correct_files_with_all_voices_dir2_filter_treble = []
for music_xml_file_path in correct_files_with_all_voices_dir2:
    score = music21.converter.parse(music_xml_file_path)
    # Get the number of streams in the parsed score
    num_streams = len(score.getElementsByClass(music21.stream.Stream))
    #print("Number of streams in the parsed score:", num_streams)
    #Test if the number of streams is the same as the number of voices in the voices folder
    voices_folder_path = os.path.join(os.path.dirname(music_xml_file_path), 'voices')
    all_voices = os.listdir(voices_folder_path)
    if 'Tr..mp3' in all_voices:
        continue
    else:
        correct_files_with_all_voices_dir2_filter_treble.append(music_xml_file_path)

In [30]:
len(correct_files_with_all_voices_dir2_filter_treble)

283

# Results
### We obtain a total of 120 files from folder 1 and 283 files for folder 2 i.e. a total of 403 folders
## Back to more filters ...

## We now apply some more musical criteria for both the folders by the musician

#### (4) Automatically discard all the folders for which, one or more voices in the score, have not at all lyrics, when should have (some editors avoid to repeat the same lyric for all voices, assuming that the musicians will relate to the first voice to read the corresponding verse). For these cases, the audio files are instrumental generated instead of voice generated. So, it is possible to automatically discard these folders by detecting in one or more than one audio files, this instrumental generation when the file name indicate a voice channel (‘soprano’, ‘alto’, ‘tenor’ ‘bass’, ‘cantus’, ‘quintus’, ‘altus’, ‘bassus’, ‘voice#’)


## Parse directory 1

In [31]:
def preprocess_string(s):
    # Remove hyphens and blank spaces
    return s.replace('-', '').replace(' ', '')

In [32]:
#Get the same number of files with score directory 1
possible_candidates_with_wrong_lyrics_dir1 = []
for music_xml_file_path in correct_files_with_all_voices_dir1_filter_treble:
    missing_lyrics = False
    #music_xml_file_path = music_xml_file['Parsable Files']
    score = music21.converter.parse(music_xml_file_path)
    all_lyrics = []
    for part in score.parts:
        lyrics_for_part = ''
        for note in part.recurse().getElementsByClass('Note'):            
            if note.lyric:
                lyrics_for_part += note.lyric
        processed_string = preprocess_string(lyrics_for_part)
        all_lyrics.append(str(processed_string))
    #print(all_lyrics)
    missing_lyrics = [s=='' for s in all_lyrics]
    if any(missing_lyrics):            
        print(music_xml_file_path)
        possible_candidates_with_wrong_lyrics_dir1.append(music_xml_file_path)

CPDL NO REPEAT_1/Grell_Nun_sich_der_Tag/score.musicxml
CPDL NO REPEAT_1/EG344Trio/score.musicxml
CPDL NO REPEAT_1/Erk_Sonntagmorgen/score.musicxml
CPDL NO REPEAT_1/Ich_danke_dem_Herrn_Schutz/score.musicxml
CPDL NO REPEAT_1/Hane-hei/score.musicxml
CPDL NO REPEAT_1/JarvS-MySoulWith/score.musicxml
CPDL NO REPEAT_1/JarvS-BlessedBe/score.musicxml
CPDL NO REPEAT_1/For_we_are_so_preciously_loved/score.musicxml
CPDL NO REPEAT_1/Jesus_holy_undefiled/score.musicxml
CPDL NO REPEAT_1/GoodPhysicianHolden1806a/score.musicxml
CPDL NO REPEAT_1/Ich_glaube_an_einen_einigen_Gott_Schutz/score.musicxml
CPDL NO REPEAT_1/I_sing_of_a_maiden/score.musicxml
CPDL NO REPEAT_1/JarvS-LordHear/score.musicxml
CPDL NO REPEAT_1/HealingMercyHolden1806/score.musicxml
CPDL NO REPEAT_1/JarvS-Christ/score.musicxml


In [33]:
len(possible_candidates_with_wrong_lyrics_dir1)

15

In [35]:
#Remove the identified songs from the final list
for item in correct_files_with_all_voices_dir1_filter_treble:
    if item in possible_candidates_with_wrong_lyrics_dir1:
        correct_files_with_all_voices_dir1_filter_treble.remove(item)

In [36]:
len(correct_files_with_all_voices_dir1_filter_treble)

107

## Parse directory 2

In [37]:
#Get the same number of files with score directory 1
possible_candidates_with_wrong_lyrics_dir2 = []
for music_xml_file_path in correct_files_with_all_voices_dir2_filter_treble:
    missing_lyrics = False
    #music_xml_file_path = music_xml_file['Parsable Files']
    score = music21.converter.parse(music_xml_file_path)
    all_lyrics = []
    for part in score.parts:
        lyrics_for_part = ''
        for note in part.recurse().getElementsByClass('Note'):            
            if note.lyric:
                lyrics_for_part += note.lyric
        processed_string = preprocess_string(lyrics_for_part)
        all_lyrics.append(str(processed_string))
    #print(all_lyrics)
    missing_lyrics = [s=='' for s in all_lyrics]
    if any(missing_lyrics):            
        print(music_xml_file_path)
        possible_candidates_with_wrong_lyrics_dir2.append(music_xml_file_path)

CPDL NO REPEAT_2/VeniCreatorTallis1567bpr/score.musicxml
CPDL NO REPEAT_2/Nh_Schumann_Rotes_R%C3%B6sleint/score.musicxml
CPDL NO REPEAT_2/Ws-pero-av2/score.musicxml
CPDL NO REPEAT_2/Nh_Eisenhofer_Lebensgenuss/score.musicxml
CPDL NO REPEAT_2/Nh_Abt_%C3%9Cber_den_Sternen_SATB/score.musicxml
CPDL NO REPEAT_2/Nh_Gregor_Gnade/score.musicxml
CPDL NO REPEAT_2/My_song_is_love_unknown/score.musicxml
CPDL NO REPEAT_2/ValeJ-ILiftMyHeart/score.musicxml
CPDL NO REPEAT_2/WoO_74_Wanderers_Nachtlied-Rheinberger-A-Kl/score.musicxml
CPDL NO REPEAT_2/Nh_Vulpius_Nun_komm/score.musicxml
CPDL NO REPEAT_2/Nh_Staden_Gelobet/score.musicxml
CPDL NO REPEAT_2/Nh_Abt_Verlass_mich_nicht/score.musicxml
CPDL NO REPEAT_2/Of-man-and-wife-cpdl/score.musicxml
CPDL NO REPEAT_2/Nh_Loewe_Im_Fr%C3%BChling/score.musicxml
CPDL NO REPEAT_2/Nh_Drese_Seelenbr%C3%A4utigam/score.musicxml
CPDL NO REPEAT_2/Thiel_Elslein/score.musicxml
CPDL NO REPEAT_2/Nh_Fesca_Blessed/score.musicxml
CPDL NO REPEAT_2/Nh_Mendelssohn_Op19a%2C5/score.mus

In [38]:
len(possible_candidates_with_wrong_lyrics_dir2)

51

In [39]:
#Remove the identified songs from the final list
for item in correct_files_with_all_voices_dir2_filter_treble:
    if item in possible_candidates_with_wrong_lyrics_dir2:
        correct_files_with_all_voices_dir2_filter_treble.remove(item)

In [40]:
len(correct_files_with_all_voices_dir2_filter_treble)

239

In [42]:
# Save the files in csv 
#Save the filtered data based on lyrics in a csv
file_path_CPDL1 = 'correct_file_names_removed_treble_scores_cpdl_repeat1.csv'
file_path_CPDL2 = 'correct_file_names_removed_treble_scores_cpdl_repeat2.csv'
# Open the file in write mode
with open(file_path_CPDL1, 'w', newline='') as file:
    writer = csv.writer(file)
    writer.writerow(['Parsable Files'])
    for file_name in correct_files_with_all_voices_dir1_filter_treble:
        writer.writerow([file_name])

# Confirm that the list has been saved
print("All data saved successfully!")

# Open the file in write mode
with open(file_path_CPDL2, 'w', newline='') as file:
    writer = csv.writer(file)
    writer.writerow(['Parsable Files'])
    for file_name in correct_files_with_all_voices_dir2_filter_treble:
        writer.writerow([file_name])

# Confirm that the list has been saved
print("All data saved successfully!")


All data saved successfully!
All data saved successfully!


# Results
### We obtain a total of 105 files from folder 1 and 239 files for folder 2 i.e. a total of 344 folders from which we chose 20 songs and mark them as accepted

# Future Work

We can add more musically informed criteria to filter out more renditions, however, since the data is not very consistent, it is hard to implement the identified criteria. The other musical criteria for consideration includes:

(5) Automatically discard all the folders that, corresponding to strophic chorale songs, have no indication of repetition (repetition bars, da capo or da segno). This could be addressed by simple math calculation knowing the pulse, the time signature, the number of measures and the number of stanzas

Comments: Its not easy to identify how many parts are meant for Repeat and so on, and how many times the score is to be repeated

(6) Automatically discard all the folders that, instead of relating each note with its corresponding syllable of the verse, put the whole verse continuously in just the first note. As a consequence of this, the voice generation consists on the melody sung by a continuous sound “ah” for all the channels. So, it is possible to automatically discard these folders by detecting in one or more than one audio files, this onomatopoeic generation 

Comments: This criteria has been implemented by using lyrics information. If the lyrics are absent in any of the score parts, we filter out that score. 

In the analysis of musicxml scores, it was found that the lyrics information is sometimes included in place of TextExpression at the beginning of the score. The lyrics were not tied to notes in that case, or the first note. We tried implementing a way to filter out using the length of TextExpression, assuming that if the lyrics are indeed included as part of it, the length of TextExpression would be more than say a specific threshold (maybe 30), however, many times the TextExpression consists of some details of the score like the composer etc. Hence, a basic thresholding based criteria does not provide sufficient information to filter based on TextExpression. We decided to go for the lyrics approach thereafter.
 